### DISCUSSION TITLE
Updating Symfony\Component\HttpFoundation breaks legacy app

### DISCUSSION THREAD

**Discussion by paddelboot** [symfony/symfony#48058] [Q&A]
Hi,

I was updating `Symfony\Component\HttpFoundation` from 4.1.0 to v4.4.48. It broke a legacy app with `Headers already sent` errors. This is not a symfony app, I merely use some symfony packages.

In my router, I always return a `Symfony\Component\HttpFoundation\Response`object, whether a route was matched or not. If there was no match, I simply do a:

        $response = new \Symfony\Component\HttpFoundation\Response;
        return $response;

Which will then be send:

        $response->prepare( $request );
        $response->send();

Can anyone explain whether in between above versions, there was a change in how headers were appended to the default instance of the `response` object? It seems that in 4.1.0, no headers were added at all, preventing the above error.

**Comment by MatTheCat** [comment]
The only change which comes to my mind is https://github.com/symfony/symfony/pull/46931 / https://github.com/symfony/symfony/pull/47434 which ensure headers **cannot** be sent anyway after the response has been `sent` :thinking: 

Given `Response` checks if the headers already have been I don’t even know how this could happen inside of Symfony. https://github.com/symfony/symfony/blob/a23b1f3be1c5a6818481f4fd10792e8400446f65/src/Symfony/Component/HttpFoundation/Response.php#L338-L340

Could you provide a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html)?

**Comment by paddelboot** [comment]
When no route matches, sometimes the app does something else, like redirecting to a login page. And this is precisely where my code breaks, since headers with `v4.4.48` are *always* sent, it seems, after issuing `$response->send();`. At least this is how I try to explain it to myself. 
If I rollback to `v4.1.0` everything works againg. Will see if I can find the time to reproduce this.

**Reply by MatTheCat** [reply]
`Response`’s headers are not sent if you already sent yours before (cf. the code I pasted in my previous message), so a reproducer will be helpful!

Also please write in “Write a reply” instead of “Suggest an answer” to reply to a message.
